Mathlib/NumberTheory/Ostrowski.lean

```diff
@@ -228,4 +258,91 @@ theorem mulRingNorm_equiv_padic_of_bounded :
+  f n = f ((Nat.ofDigits m L : ℕ) : ℚ) := by rw [Nat.ofDigits_digits m n]
+    _ = f (L'.sum) := by rw [Nat.ofDigits_eq_sum_mapIdx]; norm_cast
+    _ ≤ (L'.map f).sum := mulRingNorm_sum_le_sum_mulRingNorm L' f
+    _ ≤ (L.mapIdx fun i _ ↦ m * (f m) ^ i).sum := ?_
+  simp only [hL', List.mapIdx_eq_enum_map, List.map_map]
+  apply List.sum_le_sum
+  rintro ⟨i, a⟩ hia
+  dsimp [Function.uncurry]
+  replace hia := List.mem_enumFrom hia
+  push_cast
+  rw [map_mul, map_pow]
+  apply mul_le_mul_of_nonneg_right (le_of_lt (hcoef _))
+    (pow_nonneg (apply_nonneg f ↑m) i)
+  simp only [zero_le, zero_add, tsub_zero, true_and] at hia
+  exact List.mem_iff_get.mpr ⟨⟨i, hia.1⟩, hia.2.symm⟩
+
+/-! ## Step 1: if f is a MulRingNorm and f n > 1 for some natural n, then f n > 1 for all n ≥ 2 -/
+
+/-- If `f n > 1` for some `n` then `f n > 1` for all `n ≥ 2` -/
+lemma one_lt_of_not_bounded (notbdd : ¬ ∀ n : ℕ, f n ≤ 1) {n₀ : ℕ} (hn₀ : 1 < n₀) : 1 < f n₀ := by
+  contrapose! notbdd with h
+  intro n
+  have h_ineq1 {m : ℕ} (hm : 1 ≤ m) : f m ≤ n₀ * (logb n₀ m + 1) := by
+    /- L is the string of digits of `n` in the base `n₀`-/
+    set L := Nat.digits n₀ m
+    calc
+    f m ≤ (L.mapIdx fun i _ ↦ n₀ * f n₀ ^ i).sum := mulRingNorm_apply_le_sum_digits m hn₀
+    _ ≤ (L.mapIdx fun _ _ ↦ (n₀ : ℝ)).sum := by
+      simp only [List.mapIdx_eq_enum_map, List.map_map]
+      apply List.sum_le_sum
+      rintro ⟨i, a⟩ _
+      simp only [Function.comp_apply, Function.uncurry_apply_pair]
+      exact mul_le_of_le_of_le_one' (mod_cast le_refl n₀) (pow_le_one₀ (apply_nonneg f ↑n₀) h)
+        (pow_nonneg (apply_nonneg f ↑n₀) i) (Nat.cast_nonneg n₀)
+    _ = n₀ * (Nat.log n₀ m + 1) := by
+      rw [List.mapIdx_eq_enum_map, List.eq_replicate_of_mem (a := (n₀ : ℝ))
+        (l := List.map (Function.uncurry fun _ _ ↦ n₀) (List.enum L)),
+        List.sum_replicate, List.length_map, List.enum_length, nsmul_eq_mul, mul_comm,
+        Nat.digits_len n₀ m hn₀ (not_eq_zero_of_lt hm), Nat.cast_add_one]
+      simp (config := { contextual := true })
+    _ ≤ n₀ * (logb n₀ m + 1) := by gcongr; exact natLog_le_logb ..
+  -- For h_ineq2 we need to exclude the case n = 0.
+  rcases eq_or_ne n 0 with rfl | h₀
+  · simp only [CharP.cast_eq_zero, map_zero, zero_le_one]
+  have h_ineq2 (k : ℕ) (hk : 0 < k) :
+      f n ≤ (n₀ * (logb n₀ n + 1)) ^ (k : ℝ)⁻¹ * k ^ (k : ℝ)⁻¹ := by
+    have h_exp : (f n ^ (k : ℝ)) ^ (k : ℝ)⁻¹ = f n := by
+      apply rpow_rpow_inv (apply_nonneg f ↑n)
+      simp only [ne_eq, Nat.cast_eq_zero]
+      exact Nat.pos_iff_ne_zero.mp hk
+    have : 0 ≤ logb n₀ n := logb_nonneg (one_lt_cast.2 hn₀) (mod_cast Nat.one_le_of_lt h₀.bot_lt)
+    rw [← mul_rpow (by positivity) (by positivity), ← h_exp, rpow_natCast, ← map_pow,
+      ← Nat.cast_pow]
+    gcongr
+    apply le_trans (h_ineq1 (one_le_pow k n h₀.bot_lt))
+    rw [mul_assoc, Nat.cast_pow, logb_pow (mod_cast h₀.bot_lt), mul_comm _ (k : ℝ),
+      mul_add (k : ℝ), mul_one]
+    gcongr
+    exact one_le_cast.mpr hk
```

Comment:
I feel like this proof could be made into a nice `calc` block. Mind giving it a go?

Reply:
Done